File: node/zkgroup/ServerSecretParams.ts
import { randomBytes } from 'crypto';
import NativeImpl from '../NativeImpl';
import ByteArray from './internal/ByteArray';

import { RANDOM_LENGTH } from './internal/Constants';
import ServerPublicParams from './ServerPublicParams';
import NotarySignature from './NotarySignature';

export default class ServerSecretParams extends ByteArray {

  static SIZE = 1121;

  static generate(): ServerSecretParams {
    const random = randomBytes(RANDOM_LENGTH);

    return ServerSecretParams.generateWithRandom(random);
  }

  static generateWithRandom(random: Buffer): ServerSecretParams {
    return new ServerSecretParams(NativeImpl.ServerSecretParams_GenerateDeterministic(random));
  }

  constructor(contents: Buffer)  {
    super(contents, ServerSecretParams.SIZE, true);
    NativeImpl.ServerSecretParams_CheckValidContents(contents);
  }

  getPublicParams(): ServerPublicParams {
    return new ServerPublicParams(NativeImpl.ServerSecretParams_GetPublicParams(this.contents));
  }

  sign(message: Buffer): NotarySignature {
    const random = randomBytes(RANDOM_LENGTH);

    return this.signWithRandom(random, message);
  }

  signWithRandom(random: Buffer, message: Buffer): NotarySignature {
    return new NotarySignature(NativeImpl.ServerSecretParams_SignDeterministic(this.contents, random, message));
  }
}

File: java/java/src/main/java/org/signal/zkgroup/ServerSecretParams.java
//
// Copyright 2020-2021 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

package org.signal.zkgroup;

import java.security.SecureRandom;
import org.signal.zkgroup.internal.ByteArray;
import org.signal.client.internal.Native;

import static org.signal.zkgroup.internal.Constants.RANDOM_LENGTH;

public final class ServerSecretParams extends ByteArray {

  public static ServerSecretParams generate() {
    return generate(new SecureRandom());
  }

  public static ServerSecretParams generate(SecureRandom secureRandom) {
    byte[] random      = new byte[RANDOM_LENGTH];
    secureRandom.nextBytes(random);

    byte[] newContents = Native.ServerSecretParams_GenerateDeterministic(random);

    try {
      return new ServerSecretParams(newContents);
    } catch (IllegalArgumentException e) {
      throw new AssertionError(e);
    } 
  }

  public ServerSecretParams(byte[] contents)  {
    super(contents);
    Native.ServerSecretParams_CheckValidContents(contents);
  }

  public ServerPublicParams getPublicParams() {
    byte[] newContents = Native.ServerSecretParams_GetPublicParams(contents);
    return new ServerPublicParams(newContents);
  }

  public NotarySignature sign(byte[] message) {
    return sign(new SecureRandom(), message);
  }

  public NotarySignature sign(SecureRandom secureRandom, byte[] message) {
    byte[] random      = new byte[RANDOM_LENGTH];
    secureRandom.nextBytes(random);

    byte[] newContents = Native.ServerSecretParams_SignDeterministic(contents, random, message);

    try {
      return new NotarySignature(newContents);
    } catch (InvalidInputException e) {
      throw new AssertionError(e);
    }
  }

}
